django-nonrel/ocl/oclapi/models.py

```diff
@@ -37,12 +41,29 @@ class BaseModel(models.Model):
     created_by = models.TextField()
     updated_by = models.TextField()
     is_active = models.BooleanField(default=True)
+    is_being_saved = False
     extras = DictField(null=True, blank=True)
+    extras_are_encoded = False
+    extras_are_decoded = False
     uri = models.TextField(null=True, blank=True)
 
     class Meta:
         abstract = True
 
+
+    def save(self, force_insert=False, force_update=False, using=None, update_fields=None):
+        self.is_being_saved = True
```

Comment:
I understand you introduced this variable so that when encode_extras is called __setattr__ will not decode it at the same time. I don't think it is needed. Instead I would just call super(BaseModel, self).__setattr__('extras', encoded_extras) in encode_extras to avoid it being decoded at the same time.

Reply:
Since the values are decoded in the setter (and the setter is called very many times from very many places including during saves, fetches, views, etc.) I believe this is the safest way to ensure values are correctly encoded/decoded.  Adding this boolean ensures that values are not decoded by setattr (which is called multiple times by the framework that does the save operation) during the save operation performed by the framework.  The method shown below is called multiple times during the save operation performed by “super(BaseModel, self).save(force_insert, force_update, using, update_fields)”

```
def __setattr__(self, attrname, val):
    if("extras" == attrname and val is not None and self.is_being_saved == False and self.extras_are_decoded == False):
        val = self.decode_extras(val)
        self.extras_are_decoded = True
    super(BaseModel, self).__setattr__(attrname, val)

```

Reply:
Okay, please write some tests to prove your solution works and I will accept it based on that.